In [ ]:
#=
Consider the following system of 2 asymmetrically coupled pendula. 

  +--------+--------+
  |                 |
  +---||---+---||---+
      lₐ       lᵦ   
      ||       ||   
      m₁---k---m₂

Where:
m₁: mass of left pendulem
m₂: mass of right pendulem
lₐ: length of left pendulem
lᵦ: length of right pendulem
k: spring constant
=#

Using Newton's second law, we have 2 equations of motion for the 2 mass system:
$$
\begin{aligned}
m_1 \ddot{x_1} = -k(x_1 - x_2) - \frac{m_1 g}{l_a}x_1 \\
m_2 \ddot{x_2} = -k(x_2 - x_1) - \frac{m_2 g}{l_{\beta}}x_2
\end{aligned} \Longrightarrow \begin{aligned}
m_1 \ddot{x_1} = - kx_1 + kx_2 - \frac{m_1 g}{l_a}x_1 \\
m_2 \ddot{x_2} = - kx_2 + kx_1 - \frac{m_2 g}{l_{\beta}}x_2
\end{aligned} \Longrightarrow \begin{aligned}
\ddot{x_1} = - \frac{k}{m_1}x_1 + \frac{k}{m_1}x_2 - \frac{g}{l_a}x_1 \\
\ddot{x_2} = - \frac{k}{m_2}x_2 + \frac{k}{m_2}x_1 - \frac{g}{l_{\beta}}x_2
\end{aligned} 
$$
$$
\begin{aligned}
\ddot{x_1} = \left(- \frac{k}{m_1} - \frac{g}{l_a}\right)x_1 + \left(\frac{k}{m_1}\right)x_2 \\
\ddot{x_2} = \left(\frac{k}{m_2}\right)x_1 + \left(- \frac{k}{m_2} - \frac{g}{l_{\beta}}\right)x_2
\end{aligned} 
$$
We make the ansatz $ x_j = x_j^0 e^{i \omega t}$, where mass index $j = (1,2)$ since we have two masses and the $^0$ superscript indicates initial position. Differentiating, 
$$
\begin{aligned}
x_j &= x_j^0 e^{i \omega t} \\
\dot{x_j} &= i \omega x_j^0 e^{i \omega t} \\
\ddot{x_j}&= -\omega^2 x_j^0 e^{i \omega t}
\end{aligned}
$$
Substituting into the equation motions,

$$ 
\begin{aligned}
-\omega^2 x_1^0 e^{i \omega t} &= \left(- \frac{k}{m_1} - \frac{g}{l_a}\right)x_1^0 e^{i \omega t} + \left(\frac{k}{m_1}\right)x_2^0 e^{i \omega t} \\
-\omega^2 x_2^0 e^{i \omega t} &= \left(\frac{k}{m_2}\right) x_1^0 e^{i \omega t} + \left(- \frac{k}{m_2} - \frac{g}{l_{\beta}}\right)x_2^0 e^{i \omega t} 
\end{aligned}
$$

This leads to:

$$
\begin{aligned}
-\omega^2 x_1^0 &= \left(- \frac{k}{m_1} - \frac{g}{l_a}\right)x_1^0  + \left(\frac{k}{m_1}\right)x_2^0 \\
-\omega^2 x_2^0 &= \left(\frac{k}{m_2}\right)x_1^0  + \left(- \frac{k}{m_2} - \frac{g}{l_{\beta}}\right)x_2^0 
\end{aligned}
$$

Casting the system into matrix form,

$$
-\omega^2 \vec{x} = \begin{bmatrix}
- \frac{k}{m_1} - \frac{g}{l_a} & \frac{k}{m_1} \\
\frac{k}{m_2} & - \frac{k}{m_2} - \frac{g}{l_{\beta}}
\end{bmatrix} \vec{x}
$$

With numerical values, we can calculate the eigenvalues and corresponding eigenvectors for this matrix. Let $m_1 = 1$ kg, $m_2 = 2$ kg, $l_a = 1$ m, $ l_\beta = 1 $ m, $ k = 1 $ N/m, and $ g = 10 $ m/s².


In [7]:
using LinearAlgebra

m₁ = 1
m₂ = 2
lₐ = 1
lᵦ = 1
k = 1
g = 10

A11 = - (k / m₁) - (g / lₐ)
A12 = k / m₁
A21 = k / m₂
A22 = - (k / m₂) - (g / lᵦ)

A = [A11 A12
     A21 A22]

display(A)
eigenvalues, eigenvectors = eigen(A)

for i in 1:2
    println("NORMAL MODE ", i)
    println("Eigenvalue: ", round(eigenvalues[i], digits = 3))
    println("Eigenvector: ", round.(eigenvectors[:, i], digits = 3))
    println()
end

2×2 Matrix{Float64}:
 -11.0    1.0
   0.5  -10.5

NORMAL MODE 1
Eigenvalue: -11.5


Eigenvector: [-0.894, 0.447]

NORMAL MODE 2
Eigenvalue: -10.0
Eigenvector: [-0.707, -0.707]



Other symbolic algebra programs might output $|e_2\rangle$ = [0.707, 0.707], which represents the same eigendirection.